nixos: systemd: split module up into multiple files

bobvanderlinden opened the pull request:
###### Description of changes

Currently the systemd module is quite large. It contains multiple components that could be split off into separate files.

The idea behind this refactor is to not only make the modules more readable, but it is also a step towards more reusable systemd configuration building.

This PR does introduce a new option called `systemd.additionalUpstreamUserUnits`. It was introduced to have systemd-tmpfiles specify its own user units, instead of systemd-user doing that for systemd-tmpfiles. I think the way these upstream units still needs a bit of work, but I intend to do that in a separate PR.

A next step is to generalize unit directory building between systemd system and systemd user. The main intention is to also reuse this for systemd in initrd (https://github.com/NixOS/nixpkgs/pull/120015). It avoids creating yet another system of unit building.

I did not ran all tests, but I did ran the systemd tests as well as applied this configuration to my own system.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

bobvanderlinden committed:
nixos: systemd: split off logind into separate module

bobvanderlinden committed:
nixos: systemd: split off journald into separate module

bobvanderlinden committed:
nixos: systemd: split off coredump into separate module

bobvanderlinden committed:
nixos: systemd: avoid using "with systemdUtils.lib"

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/777

Artturin commented on nixos/modules/module-list.nix:
```diff
@@ -1163,7 +1163,12 @@
   ./system/boot/stage-1.nix
   ./system/boot/stage-2.nix
   ./system/boot/systemd.nix
+  ./system/boot/systemd-coredump.nix
```
```suggestion
  ./system/boot/systemd/coredump.nix
```

Imo would be nicer like this

cole-h replied:
I agree -- if we're splitting systemd into related modules, they should all go under an umbrella `systemd` directory.

bobvanderlinden replied:
I like this one as well. My previous approach aligned with `systemd-nspawn`. I have moved `systemd-nspawn` to the `systemd` directory as well in a separate commit. :+1:

SuperSandro2000 commented on nixos/modules/system/boot/systemd-user.nix:
```diff
@@ -0,0 +1,172 @@
+    pathToUnit
+    ;
+
+  upstreamUserUnits =
+    [
+      "app.slice"
+      "background.slice"
+      "basic.target"
+      "bluetooth.target"
+      "default.target"
+      "exit.target"
+      "graphical-session-pre.target"
+      "graphical-session.target"
+      "paths.target"
+      "printer.target"
+      "session.slice"
+      "shutdown.target"
+      "smartcard.target"
+      "sockets.target"
+      "sound.target"
+      "systemd-exit.service"
+      "timers.target"
+      "xdg-desktop-autostart.target"
+    ]
+    ++ config.systemd.additionalUpstreamUserUnits;
```
That isn't the formatting that was used before. Lets keep that and only split things

https://github.com/NixOS/nixpkgs/pull/164016/files#diff-5ea4693beb4bdd8f7efcdd3204ceea67c86f59b3d1c28cde063ddf03e1d20fbfL188-L210

bobvanderlinden replied:
I redid the commits to align them with the original formatting again :+1:

SuperSandro2000 commented on nixos/modules/system/boot/systemd-journald.nix:
```diff
@@ -0,0 +1,143 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
```
```suggestion
{ config, lib, pkgs, ... }:

```
I would suggest to keep the style of the formatting like it was before. This is also nixpkgs-fmt conform. win win.

bobvanderlinden replied:
Done :+1:

bobvanderlinden committed:
nixos: systemd: split off systemd-user into separate module

bobvanderlinden committed:
nixos: systemd: split off systemd-tmpfiles into separate module

bobvanderlinden committed:
nixos: systemd-tmpfiles: use cfg shorthand

bobvanderlinden committed:
nixos: systemd-journald: use cfg shorthand

bobvanderlinden committed:
nixos: systemd-user: use cfg shorthand

bobvanderlinden committed:
nixos: systemd-user: allow additional upstream user units

Currently it is only possible to add upstream _system_ units. The option
systemd.additionalUpstreamSystemUnits can be used for this.

However, this was not yet possible for systemd.user. In a similar
fashion this was added to systemd-user.nix.

This is intended to have other modules add upstream units.

bobvanderlinden committed:
nixos: systemd: move systemd-nspawn to systemd/nspawn

bobvanderlinden committed:
nixos: systemd-logind: use cfg shorthand

bobvanderlinden committed:
nixos: systemd-coredump: use cfg shorthand

bobvanderlinden commented:
Unfortunately I couldn't make the build unchanged, because the order of `upstreamSystemUnits` and `upstreamUserUnits` changed (due to merging the lists). The order of those lists shouldn't affect functionality.

I ran for each of the commits to make sure they succeed:

```
nix-build nixos/release.nix -A tests.systemd.x86_64-linux
```

cole-h commented:
cc @NixOS/systemd

flokli commented on nixos/modules/system/boot/systemd/user.nix:
```diff
@@ -0,0 +1,157 @@
+      type = types.listOf types.str;
+      example = [];
+      description = ''
+        Additional user units shipped with systemd that shall be enabled.
```
For which user(s) will these be enabled? All users that have a login session?

If we trace its usages, I can see they should end up as files in `/etc/systemd/user`, alongside with all the ones in the static `upstreamUserUnits` list, but it probably shouldn't hurt to write this down, and refer to the appropriate systemd manpage.

bobvanderlinden replied:
The units will just be available to `systemd --user`. This works the same as before. I doubt any NixOS user would want to add/remove someting to `upstreamUserUnits`. Unless they are developing NixOS itself. Same goes for `upstreamSystemUnits`.

I'm thinking it might be best to add `internal = true;` and possibly `visible = false;` to this option.

bobvanderlinden replied:
I've opted for expanding on the description a bit more to mention per-user systemd instances and also make the option internal.

ElvishJerricco replied:
@flokli I think that's tangential to this PR, given this PR is just moving around existing code.

bobvanderlinden committed:
nixos: systemd-user: expand on additionalUpstreamUserUnits description

bobvanderlinden committed:
nixos: systemd-user: make additionalUpstreamUserUnits internal

ElvishJerricco commented:
I would really like to see this merge. It's been helpful in my systemd-initramfs effort

arianvp commented:
@GrahamcOfBorg test systemd

arianvp commented:
Could we squash the `systemd-user` bits into a single commit or split it off into a new PR?  Given it adds new functionaltiy it's a bit hard to follow that it's interwoven with plain refactorings.

Apart from that LGTM

dasJ commented on nixos/modules/system/boot/systemd/coredump.nix:
```diff
@@ -0,0 +1,58 @@
+  cfg = config.systemd.coredump;
+  systemd = config.systemd.package;
+in {
+  options = {
+    systemd.coredump.enable = mkOption {
```
```suggestion
  options.systemd.coredump = {
    enable = mkOption {
```

dasJ replied:
Debatable change but if done should also done to the other modules

bobvanderlinden replied:
Meh, most files followed the `options = {` style approach, so I followed by example. It's fine to change them all and I'm fine to do that, but it seems more fitting in a separate PR.

If there is an unwritten rule that `options.* =` style approach should be taken, let's write that down somewhere. (linting rules?)

dasJ commented on nixos/modules/system/boot/systemd/coredump.nix:
```diff
@@ -0,0 +1,58 @@
+      '';
+    };
+
+    systemd.coredump.extraConfig = mkOption {
```
```suggestion
    extraConfig = mkOption {
```

bobvanderlinden replied:
Same as above. This conflicts with @SuperSandro2000's comment to keep the code as-is in this PR as much as possible to help the review process.

dasJ commented on nixos/modules/system/boot/systemd/coredump.nix:
```diff
@@ -0,0 +1,58 @@
+
+  config = {
+    systemd.additionalUpstreamSystemUnits = [
+      # Coredumps.
```
```suggestion

```

dasJ commented on nixos/modules/system/boot/systemd/journald.nix:
```diff
@@ -0,0 +1,132 @@
+
+  config = {
+    systemd.additionalUpstreamSystemUnits = [
+      # Journal.
```
```suggestion

```

dasJ commented on nixos/modules/system/boot/systemd/logind.nix:
```diff
@@ -0,0 +1,115 @@
+
+  config = {
+    systemd.additionalUpstreamSystemUnits = [
+      # Login stuff.
```
```suggestion

```

dasJ commented on nixos/modules/system/boot/systemd/tmpfiles.nix:
```diff
@@ -0,0 +1,105 @@
+
+  config = {
+    systemd.additionalUpstreamSystemUnits = [
+      # Temporary file creation / cleanup.
```
```suggestion

```

dasJ reviewed (commented):
Changes look great, apart from my (somewhat opinionated) minor style nits ;)

I'd be happy if this went through without my concerns being adressed, so I'm only commenting here and not requesting changes.

bobvanderlinden committed:
nixos: systemd: remove unhelpful comments in additionalUpstreamSystemUnits

Co-authored-by: Janne Heß <janne@hess.ooo>

bobvanderlinden commented:
> I would really like to see this merge. It's been helpful in my systemd-initramfs effort

Review approval would help :smile: :+1:

dasJ reviewed (approved):
LGTM. If nobody brings up arguments why I shouldn't, I'd merge this tomorrow or so

mweinelt commented on nixos/modules/system/boot/systemd/user.nix:
```diff
@@ -0,0 +1,158 @@
+      type = types.listOf types.str;
+      example = [];
+      description = ''
+        Additional units shipped with systemd that shall be enabled for per-user systemd instances.
```
```suggestion
        Additional units shipped with systemd that should be enabled for per-user systemd instances.
```

I think should is more fitting, since you're not proposing anything here.

mweinelt reviewed (approved):
Looks sane enough. Tested on my desktop and I didn't notice any regression.

bobvanderlinden committed:
nixos: systemd-user: fix additionalUpstreamUserUnits description

Co-authored-by: Martin Weinelt <mweinelt@users.noreply.github.com>

bobvanderlinden commented:
🥳 thank you all!
